geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java

```diff
@@ -88,204 +92,176 @@ public void tearDown() throws Exception {
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
     }
+    assertThatThrownBy(() -> txManager.commit()).isExactlyInstanceOf(CommitConflictException.class);
+    allowReadTransactionToProceedLatch.countDown();
   }
 
-  /**
-   * Test that four transactions with only mixed operations and read operations on same key
-   * don't produce CommitConflictException
-   * This test fill some initial value on startup and after that create different threads with
-   * region#get operations
-   * Sync two different threads commit time through lock
-   */
-  @Test(/* no exception expected */)
-  public void testDetectReadConflict4RR()
-      throws CacheException, ExecutionException, InterruptedException {
+  @Test
+  public void readConflictsTransactionCanDetectStateChange() throws Exception {
     cache.close();
     createCache();
-    TXManagerImpl txMgrImpl = (TXManagerImpl) cache.getCacheTransactionManager();
-    // fill some initial key-value on start
-    txMgrImpl.begin();
-    this.region.put("key1", "value1"); // non-tx
-    txMgrImpl.commit();
-    ExecutorService executor = Executors.newFixedThreadPool(4);
-    for (int i = 0; i < 20; i++) {
-      final StampedLock lock = new StampedLock();
-      lock.asWriteLock().lock();
-      //
-      Future<?> future1 = executor.submit(() -> {
-        CacheTransactionManager txMgr2 = cache.getCacheTransactionManager();
-        txMgr2.begin();
-        region.get("key1"); // non-tx
-        region.put("key2", "value2");
-        lock.asReadLock().lock();
-        txMgr2.commit();
-      });
-      Future<?> future2 = executor.submit(() -> {
-        CacheTransactionManager txMgr3 = cache.getCacheTransactionManager();
-        txMgr3.begin();
-        region.get("key1"); // non-tx
-        region.put("key3", "value3");
-        lock.asReadLock().lock();
-        txMgr3.commit();
-      });
-      Future<?> future3 = executor.submit(() -> {
-        CacheTransactionManager txMgr4 = cache.getCacheTransactionManager();
-        txMgr4.begin();
-        region.get("key1"); // non-tx
-        region.put("key4", "value4");
-        lock.asReadLock().lock();
-        txMgr4.commit();
-      });
-      Future<?> future4 = executor.submit(() -> {
-        CacheTransactionManager txMgr5 = cache.getCacheTransactionManager();
-        txMgr5.begin();
-        region.get("key1"); // non-tx
-        lock.asReadLock().lock();
-        txMgr5.commit();
-      });
-      pause(100);
-      lock.asWriteLock().unlock();
-      future1.get();
-      future2.get();
-      future3.get();
-      future4.get();
+
+    region.put(key, value);
+    region.put(key1, value1);
+    TXManagerImpl txManager = (TXManagerImpl) cache.getCacheTransactionManager();
+    txManager.begin();
+    assertThat(region.get(key)).isSameAs(value);
+    region.put(key1, newValue1);
+    Runnable task = () -> doPutTransaction();
+    new Thread(task).start();
```

Comment:
This is a place where you should use the ExecutorServiceRule. The Rule will ensure that 1) the task terminates and doesn't pollute later tests, 2) anything thrown by the task is correctly reported as an error, causing the test to fail.

Add this near the top of the test class:
```
@Rule
public ExecutorServiceRule executorServiceRule = new ExecutorServiceRule();
```
And then replace this line with:
```
executorServiceRule.submit(() -> doPutTransaction());
```